MSBuild treats unqualified metadata case sensitive

### Steps to reproduce

Project file
```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003" >
  <ItemGroup>
    <Foo Include="Foo">
      <SENSITIVE>X</SENSITIVE>
    </Foo>
  </ItemGroup>
  <Target Name="Build">
    
    <Message Importance="high" Text="%(Foo.FileName)=%(Foo.sensitive)"/>
    <!-- Output: Foo=X -->
    
    <Message Importance="high" Text="%(Foo.FileName)=%(sensitive)"/>
    <!-- error MSB4096: The item "Foo" in item list "Foo" does not define a value for metadata "sensitive".  
                        In order to use this metadata, either qualify it by specifying %(Foo.sensitive), or ensure that
                        all items in this list define a value for this metadata. -->
  </Target>
</Project>
```

Command line
```
msbuild
```
### Expected  behavior
Outputs `Foo=X` twice

### Actual behavior
Fails with MSB4096

### Environment data
`msbuild /version` output: 15.5.180.51428. Also tested with MSBuild from .NET 4.0.

OS info: Windows 10 1703

--- a/src/Build.UnitTests/Evaluation/Expander_Tests.cs
+++ b/src/Build.UnitTests/Evaluation/Expander_Tests.cs
@@ -30,7 +30,6 @@ using ProjectItemInstanceFactory = Microsoft.Build.Execution.ProjectItemInstance
 
 using Xunit;
 using Microsoft.Build.BackEnd;
-using Microsoft.Build.Engine.UnitTests;
 using Microsoft.Build.Shared.FileSystem;
 using Shouldly;
 
@@ -1136,6 +1135,85 @@ namespace Microsoft.Build.UnitTests.Evaluation
 </Project>");
 
             logger.AssertLogContains("[One|Three|Four]");
+        }
+
+        [Fact]
+        public void DirectItemMetadataReferenceShouldBeCaseInsensitive()
+        {
+            MockLogger logger = Helpers.BuildProjectWithNewOMExpectSuccess(@"
+                <Project>
+                  <ItemGroup>
+                    <Foo Include=`Foo`>
+                      <SENSITIVE>X</SENSITIVE>
+                    </Foo>
+                  </ItemGroup>
+                  <Target Name=`Build`>
+                    <Message Importance=`high` Text=`QualifiedNotMatchCase %(Foo.FileName)=%(Foo.sensitive)`/>
+                    <Message Importance=`high` Text=`QualifiedMatchCase %(Foo.FileName)=%(Foo.SENSITIVE)`/>
+                    
+                    <Message Importance=`high` Text=`UnqualifiedNotMatchCase %(Foo.FileName)=%(sensitive)`/>
+                    <Message Importance=`high` Text=`UnqualifiedMatchCase %(Foo.FileName)=%(SENSITIVE)`/>
+                  </Target>
+                </Project>
+                ");
+
+            logger.AssertLogContains("QualifiedNotMatchCase Foo=X");
+            logger.AssertLogContains("QualifiedMatchCase Foo=X");
+            logger.AssertLogContains("UnqualifiedNotMatchCase Foo=X");
+            logger.AssertLogContains("UnqualifiedMatchCase Foo=X");
+        }
+
+        [Fact]
+        public void ItemDefinitionGroupMetadataReferenceShouldBeCaseInsensitive()
+        {
+            MockLogger logger = Helpers.BuildProjectWithNewOMExpectSuccess(@"
+                <Project>
+                  <ItemDefinitionGroup>
+                    <Foo>
+                        <SENSITIVE>X</SENSITIVE>
+                    </Foo>
+                  </ItemDefinitionGroup>
+                  <ItemGroup>
+                    <Foo Include=`Foo`/>
+                  </ItemGroup>
+                  <Target Name=`Build`>
+                    <Message Importance=`high` Text=`QualifiedNotMatchCase %(Foo.FileName)=%(Foo.sensitive)`/>
+                    <Message Importance=`high` Text=`QualifiedMatchCase %(Foo.FileName)=%(Foo.SENSITIVE)`/>
+                    
+                    <Message Importance=`high` Text=`UnqualifiedNotMatchCase %(Foo.FileName)=%(sensitive)`/>
+                    <Message Importance=`high` Text=`UnqualifiedMatchCase %(Foo.FileName)=%(SENSITIVE)`/>
+                  </Target>
+                </Project>
+                ");
+
+            logger.AssertLogContains("QualifiedNotMatchCase Foo=X");
+            logger.AssertLogContains("QualifiedMatchCase Foo=X");
+            logger.AssertLogContains("UnqualifiedNotMatchCase Foo=X");
+            logger.AssertLogContains("UnqualifiedMatchCase Foo=X");
+        }
+
+        [Fact]
+        public void WellKnownMetadataReferenceShouldBeCaseInsensitive()
+        {
+            MockLogger logger = Helpers.BuildProjectWithNewOMExpectSuccess(@"
+                <Project>
+                  <ItemGroup>
+                    <Foo Include=`Foo`/>
+                  </ItemGroup>
+                  <Target Name=`Build`>
+                    <Message Importance=`high` Text=`QualifiedNotMatchCase %(Foo.Identity)=%(Foo.FILENAME)`/>
+                    <Message Importance=`high` Text=`QualifiedMatchCase %(Foo.Identity)=%(Foo.FileName)`/>
+                    
+                    <Message Importance=`high` Text=`UnqualifiedNotMatchCase %(Foo.Identity)=%(FILENAME)`/>
+                    <Message Importance=`high` Text=`UnqualifiedMatchCase %(Foo.Identity)=%(FileName)`/>
+                  </Target>
+                </Project>
+                ");
+
+            logger.AssertLogContains("QualifiedNotMatchCase Foo=Foo");
+            logger.AssertLogContains("QualifiedMatchCase Foo=Foo");
+            logger.AssertLogContains("UnqualifiedNotMatchCase Foo=Foo");
+            logger.AssertLogContains("UnqualifiedMatchCase Foo=Foo");
         }
 
         /// <summary>

--- a/src/Build/Collections/CopyOnWritePropertyDictionary.cs
+++ b/src/Build/Collections/CopyOnWritePropertyDictionary.cs
@@ -299,7 +299,7 @@ namespace Microsoft.Build.Collections
         /// </summary>
         bool IDictionary<string, T>.ContainsKey(string key)
         {
-            return PropertyNames.Contains(key);
+            return _properties.ContainsKey(key);
         }
 
         /// <summary>

